Introduce NOCACHE flag and force MD devices to pass
NOCACHE reads directly to the disk instead of handling
from cache. This is needed for MMP on MD RAID devices.
Not needed for DMU.

Index: linux-2.6.32-358.6.2.el6.x86_64/include/linux/blk_types.h
===================================================================
--- linux-2.6.32-358.6.2.el6.x86_64/include/linux/blk_types.h	2013-05-14 21:06:39.000000000 +0200
+++ linux-2.6.32-358.6.2.el6.x86_64/include/linux/blk_types.h	2013-06-06 13:16:04.040918725 +0200
@@ -128,6 +128,7 @@
 	__REQ_MIXED_MERGE,	/* merge of different types, fail separately */
 	__REQ_FLUSH,		/* request for cache flush */
 	__REQ_FLUSH_SEQ,	/* request for flush sequence */
+	__REQ_NOCACHE,		/* request bypass any cache */
 	__REQ_NR_BITS,		/* stops here */
 };
 
@@ -174,6 +175,7 @@
 #define REQ_FLUSH_SEQ		(1 << __REQ_FLUSH_SEQ)
 #define REQ_IO_STAT		(1 << __REQ_IO_STAT)
 #define REQ_MIXED_MERGE		(1 << __REQ_MIXED_MERGE)
+#define REQ_NOCACHE		(1 << __REQ_NOCACHE)
 
 #define REQ_WRITE_FLUSH		(REQ_WRITE | REQ_SYNC | REQ_NOIDLE | REQ_FLUSH)
 
Index: linux-2.6.32-358.6.2.el6.x86_64/drivers/md/raid5.c
===================================================================
--- linux-2.6.32-358.6.2.el6.x86_64/drivers/md/raid5.c	2013-05-14 21:08:11.000000000 +0200
+++ linux-2.6.32-358.6.2.el6.x86_64/drivers/md/raid5.c	2013-06-11 12:06:12.187622466 +0200
@@ -2385,6 +2385,12 @@
 		}
 		if (sector >= sh->dev[dd_idx].sector + STRIPE_SECTORS)
 			set_bit(R5_OVERWRITE, &sh->dev[dd_idx].flags);
+	} else if (bi->bi_rw & REQ_NOCACHE) {
+		/* force read from underlying device */
+		printk(KERN_WARNING "md/raid:%s: request with REQ_NOCACHE/%ld "
+		       "for stripe index %d\n", mdname(conf->mddev), bi->bi_rw,
+		       dd_idx);
+		clear_bit(R5_UPTODATE, &sh->dev[dd_idx].flags);
 	}
 	spin_unlock_irq(&conf->device_lock);
 	spin_unlock(&sh->lock);
